File: verifier-problems/src/main/java/com/jetbrains/pluginverifier/location/ProblemLocation.java
package com.jetbrains.pluginverifier.location;

import com.google.common.base.Preconditions;
import com.jetbrains.pluginverifier.persistence.Jsonable;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.TestOnly;
import org.objectweb.asm.tree.MethodNode;

/**
 * @author Sergey Patrikeev
 */
public abstract class ProblemLocation implements Jsonable<ProblemLocation> {

  //TODO: add more detailed location, e.g. superclass, field of a class, interface, throws list and so on

  @NotNull
  public static ProblemLocation fromPlugin(@NotNull String pluginId) {
    return new PluginLocation(pluginId);
  }

  @NotNull
  public static ProblemLocation fromClass(@NotNull String className) {
    return new CodeLocation(className, null, null);
  }

  @NotNull
  public static ProblemLocation fromField(@NotNull String className, @NotNull String fieldName) {
    return new CodeLocation(className, null, fieldName);
  }

  @NotNull
  @TestOnly
  public static ProblemLocation fromMethod(@NotNull String className, @NotNull String methodDescr) {
    return new CodeLocation(className, methodDescr, null);
  }

  @NotNull
  public static ProblemLocation fromMethod(@NotNull String className, @NotNull MethodNode methodNode) {
    return new CodeLocation(className, getMethodDescr(methodNode), null);
  }

  @NotNull
  private static String getMethodDescr(@NotNull MethodNode methodNode) {
    Preconditions.checkArgument(methodNode.name != null);
    Preconditions.checkArgument(methodNode.desc != null);
    return methodNode.name + methodNode.desc;
  }

  public abstract String asString();

  @Override
  public final String toString() {
    return asString();
  }

  @Override
  public final boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || o.getClass() != getClass()) return false;
    return asString().equals(((ProblemLocation) o).asString());
  }

  @Override
  public int hashCode() {
    return asString().hashCode();
  }
}

File: verifier-results/src/main/java/com/jetbrains/pluginverifier/location/ProblemLocation.kt
package com.jetbrains.pluginverifier.location

import org.jetbrains.annotations.TestOnly
import org.objectweb.asm.tree.MethodNode

/**
 * @author Sergey Patrikeev
 */
abstract class ProblemLocation {

  abstract fun asString(): String

  final override fun toString(): String = asString()

  final override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null || other.javaClass != javaClass) return false
    return asString() == (other as ProblemLocation).asString()
  }

  final override fun hashCode(): Int = asString().hashCode()

  companion object {

    //TODO: add more detailed location, e.g. superclass, field of a class, interface, throws list and so on

    fun fromPlugin(pluginId: String): ProblemLocation = PluginLocation(pluginId)

    fun fromClass(className: String): ProblemLocation = CodeLocation(className, null, null)

    fun fromField(className: String, fieldName: String): ProblemLocation = CodeLocation(className, null, fieldName)

    @TestOnly
    fun fromMethod(className: String, methodDescr: String): ProblemLocation = CodeLocation(className, methodDescr, null)

    fun fromMethod(className: String, methodNode: MethodNode): ProblemLocation = CodeLocation(className, getMethodDescr(methodNode), null)

    private fun getMethodDescr(methodNode: MethodNode): String = methodNode.name + methodNode.desc
  }
}
